### PR TITLE
Version 4

### PR DESCRIPTION
**Pull request by StorytellerCZ**
* BREAKING: Synchronous functions are now available only on client side
* BREAKING: Migration functions for past versions are no longer available
* Removed deprecated publication of legacy role assignments
* UI Helpers now use `console.debug` for debugging messages
* `RolesCollection` and `RoleAssignmentsCollection` can now be exported in addition to being accessed via `Meteor.roles` and `Meteor.roleAssignment`
* Support for Meteor 3

### CODE REVIEW

**Comment by jankapunkt** [comment]
@StorytellerCZ is there anything that isn't already implemented by the latest release that contains the async counterparts?

**Comment by StorytellerCZ** [comment]
Not 100% sure, but I think we have everything. I will remove the sync MongoDB calls in this branch. Or maybe for the client we can leave the sync methods? Thoughts?

**Comment by tchax** [comment]
@StorytellerCZ I think it would make sense to leave sync methods for the client. This is consistent with other packages like Accounts isn't it ? Thanks

**Comment by dallman2** [comment]
Synchronous Mongo calls on the client side are alright in Meteor 3.0, IIRC. When is this PR expected to be merged in and published? I would love to test this in my application.

**Comment by StorytellerCZ** [comment]
Once this is merged:
https://github.com/Meteor-Community-Packages/meteor-roles/pull/344

I will release a beta.

**Comment by dallman2** [comment]
I would suggest updating the `userIsInRole` function on like 658 of `roles_common.js` to: 

```
  userIsInRole: async function (user, roles, options) {
    let id
    options = Roles._normalizeOptions(options)

    // ensure array to simplify code
    if (!Array.isArray(roles)) roles = [roles]

    roles = roles.filter(r => r != null)

    if (!roles.length) return false

    Roles._checkScopeName(options.scope)

    options = Object.assign({
      anyScope: false
    }, options)

    if (user && typeof user === 'object') {
      id = user._id
    } else {
      id = user
    }

    if (!id) return false
    if (typeof id !== 'string') return false

    const selector = { 'user._id': id }

    if (!options.anyScope) {
      selector.scope = { $in: [options.scope, null] }
    }

    const promises = roles.map((roleName) => {
      selector['inheritedRoles._id'] = roleName

      return Meteor.roleAssignment.find(selector, { limit: 1 }).countAsync() > 0
    })
    await Promise.all(promises)
    return promises.some(p => p)
  },
```

For obvious reasons, the method needs to become async to be able to count how many records are in the collection.

**Comment by StorytellerCZ** [comment]
@dallman2 PRs welcome.

File: roles/roles_client.js
Comment (jankapunkt) [inline roles/roles_client.js:68]: This is breaking all our tests:

1. If we directly replace `Meteor.Roles` it will not apply this internal collection. We also can no longer check if a role is created using `Meteor.roles.countDocuments` 
2. If we use `hwillston:stub-collections` we can only test for integration and not manully stub specific methods for unit tests. This was possible when `Meteor.roles` was used.

Comment (StorytellerCZ) [reply roles/roles_client.js:68]: The `Meteor.roles` is still applied and unchanged, but you can now import he RolesCollection as well from the package.

Comment (StorytellerCZ) [reply roles/roles_client.js]: @jankapunkt please give this a look in the latest push, I have reverted to using the `Meteor` object for collection access.

File: package.js
Comment (jankapunkt) [inline package.js:45]: Not sure if we should explicitly use `meteortesting:mocha@3.2.0` since before this version there was a little mess with the versions

File: roles/roles_server.js
Comment (jankapunkt) [inline roles/roles_server.js:15]: shouldn't there be an `await` here?

Comment (StorytellerCZ) [reply roles/roles_server.js:15]: No. Creation of indexes is not a priority that needs to hold up the application.

File: package.js
Comment (StorytellerCZ) [reply package.js:45]: Improving!

**Comment by StorytellerCZ** [comment]
Anything else for release? @jankapunkt @harryadel

File: .github/workflows/testsuite.yml
Comment (harryadel) [inline .github/workflows/testsuite.yml:41]: What's the use of testing on older versions if we're going to introduce breaking changes? I think it'd be best to only have 3 and up.

File: testapp/.meteor/release
Comment (harryadel) [inline testapp/.meteor/release:1]: The test app should be a 3.0 application, no?

**Review by harryadel — APPROVED** [review]
Very few nitpicks, feel free to disregard if you don't agree.

File: .github/workflows/testsuite.yml
Comment (StorytellerCZ) [reply .github/workflows/testsuite.yml:41]: We are technically backward compatible to 2.8.1.

File: testapp/.meteor/release
Comment (StorytellerCZ) [reply testapp/.meteor/release:1]: I was thinking about this and in the end settled on 2.16. In theory the test suite matrix should dictate this version, but I don't think that is the case. Since we are still backward compatible to 2.8.1 maybe the best approach would be to create another test suite that is v3 app. Thoughts?

Comment (harryadel) [reply testapp/.meteor/release:1]: Yeah, I think having a dedicated 3.0 test application is the way to go, though it's going to be cumbersome maintaining two test applications at the same time, I'd recommend to double down on 3.0 but it's up to you if you can dedicate the time.  :+1:

Comment (StorytellerCZ) [reply testapp/.meteor/release:1]: Well, there are some other things in motion which do not make it a big issue. Just need to get it working properly.

Comment (jankapunkt) [reply testapp/.meteor/release:1]: If you want a test matrix in GH Actions then you need to create a new Meteor bare project each time the matrix release changes so we need to adjust the CI and use the test project only locally.

I can provide that if you need.

Comment (StorytellerCZ) [reply testapp/.meteor/release:1]: Maybe later. I don't think that the effort is worth it at the moment, given the future plans for this package. The current setup is good enough.

**Review by jankapunkt — APPROVED** [review]

**Pull request merged**
